File: 프로그래머스/2/131127. 할인 행사/할인 행사.java
import java.util.HashMap;
import java.util.Map;

public class Solution {
    public int solution(String[] want, int[] num, String[] disc) {
        // 필요한 제품 목록과 수량을 딕셔너리로 변환
        Map<String, Integer> wantMap = new HashMap<>();
        for (int i = 0; i < want.length; i++) {
            wantMap.put(want[i], num[i]); // 제품별 필요한 개수 저장
        }

        // 슬라이딩 윈도우 크기 설정 (number 배열의 총합, 항상 10)
        int winSize = 0;
        for (int n : num) {
            winSize += n;
        }

        int count = 0; // 원하는 제품을 모두 구매할 수 있는 날짜 수

        // 슬라이딩 윈도우를 활용한 비교
        for (int i = 0; i <= disc.length - winSize; i++) {
            // 현재 윈도우 내 상품 수량 계산
            Map<String, Integer> winMap = new HashMap<>();
            for (int j = 0; j < winSize; j++) {
                String item = disc[i + j]; // 현재 윈도우에서 할인되는 제품
                winMap.put(item, winMap.getOrDefault(item, 0) + 1); // 제품 개수 증가
            }

            // 필요한 제품 수량 충족 여부 확인
            boolean allSatisfied = true;
            for (String item : wantMap.keySet()) {
                if (winMap.getOrDefault(item, 0) < wantMap.get(item)) {
                    allSatisfied = false; // 필요한 개수를 충족하지 못하면 실패
                    break;
                }
            }

            if (allSatisfied) {
                count++; // 조건 만족 시 카운트 증가
            }
        }

        return count; // 원하는 제품을 모두 할인받을 수 있는 날짜 수 반환
    }

    public static void main(String[] args) {
        Solution sol = new Solution();

        // 결과 --> 3
        System.out.println(sol.solution(
            new String[]{"banana", "apple", "rice", "pork", "pot"},
            new int[]{3, 2, 2, 2, 1},
            new String[]{"chicken", "apple", "apple", "banana", "rice",
                         "apple", "pork", "banana", "pork", "rice", "pot",
                         "banana", "apple", "banana"}));

        // 결과 --> 0
        System.out.println(sol.solution(
            new String[]{"apple"},
            new int[]{10},
            new String[]{"banana", "banana", "banana", "banana", "banana",
                         "banana", "banana", "banana", "banana", "banana"}));
    }
}

File: 프로그래머스/2/131127. 할인 행사/할인 행사.js
function solution(want, number, discount) {
    // 요청 상품과 수량을 해시맵으로 구성
    const wantDict = {};
    for (let i = 0; i < want.length; i++) {
        wantDict[want[i]] = number[i];
    }

    // 슬라이딩 윈도우 크기: 요청한 상품의 총 수량 (예: 항상 10)
    const windowSize = number.reduce((acc, v) => acc + v, 0);
    let count = 0;

    // 가능한 모든 윈도우 시작 지점 탐색
    for (let i = 0; i <= discount.length - windowSize; i++) {
        // 현재 윈도우 부분 배열 추출
        const window = discount.slice(i, i + windowSize);

        // 윈도우 내 상품 수량을 해시맵으로 계산
        const windowCounter = {};
        for (const item of window) {
            if (windowCounter[item] === undefined) {
                windowCounter[item] = 1;
            } else {
                windowCounter[item]++;
            }
        }

        // 모든 요청 상품이 충분히 포함되었는지 확인
        let isValid = true;
        for (const item in wantDict) {
            if (!windowCounter[item] || windowCounter[item] < wantDict[item]) {
                isValid = false;
                break;
            }
        }

        // 조건 만족 시 count 증가
        if (isValid) {
            count++;
        }
    }

    // 결과 반환
    return count;
}

// 테스트 실행
console.log(
    solution(
        ["banana", "apple", "rice", "pork", "pot"],
        [3, 2, 2, 2, 1],
        [
            "chicken", "apple", "apple", "banana", "rice",
            "apple", "pork", "banana", "pork", "rice", "pot",
            "banana", "apple", "banana"
        ]
    )
); // 3

console.log(
    solution(
        ["apple"],
        [10],
        Array(10).fill("banana")
    )
); // 0
